github/oauth_server.go

```diff
@@ -13,82 +17,107 @@ import (
 )
 
 var (
-	ServerBind    = "127.0.0.1:7024"
-	ServerAddress = "http://127.0.0.1:7024"
-)
-
-var (
-	oauthConf = oauth2.Config{
-		ClientID:     "d5a2938d576279fbc995",
-		ClientSecret: "aadcab62248629c3d0018b9e54e4d3fc37fdd29e",
-		Scopes:       []string{"user:email", "repo"},
-		Endpoint:     githuboauth.Endpoint,
-	}
+	oauthClientId       = "d5a2938d576279fbc995"
+	oauthScopes         = "user:email,repo"
+	githubStageOneUri   = "https://github.com/login/device/code"
+	githubStageThreeUri = "https://github.com/login/oauth/access_token"
+	githubGrantType     = "urn:ietf:params:oauth:grant-type:device_code"
 )
 
 func NewClientFromToken(token *oauth2.Token) *githubclient.Client {
+	oauthConf := oauth2.Config{
+		ClientID: oauthClientId,
+		Endpoint: githuboauth.Endpoint,
+	}
 	oauthClient := oauthConf.Client(oauth2.NoContext, token)
 	return githubclient.NewClient(oauthClient)
 }
 
 func Authenticate() (*oauth2.Token, error) {
-	ln, err := net.Listen("tcp", ServerBind)
+
+	resp, err := http.Post(fmt.Sprintf("%s?client_id=%s&scope=%s", githubStageOneUri, oauthClientId, oauthScopes), "application/json", nil)
 	if err != nil {
 		return nil, err
 	}
+	defer resp.Body.Close()
 
-	handler := githubAuthHandler{
-		OAuthStateString: "da39a3ee5e6b4b0d3255bfef95601890afd80709",
-		TokenCh:          make(chan *oauth2.Token),
+	if resp.StatusCode != 200 {
+		return nil, errors.New(fmt.Sprintf("Github returned non-200 response (%d)", resp.Status))
 	}
+	body, err := ioutil.ReadAll(resp.Body)
 
-	go func() {
-		_ = http.Serve(ln, handler)
-	}()
+	fmt.Println("Response body:", string(body))
 
-	if err := open.Run(ServerAddress); err != nil {
+	params, err := url.ParseQuery(string(body))
+	if err != nil {
 		return nil, err
 	}
 
-	select {
-	case token := <-handler.TokenCh:
-		return token, ln.Close()
-	case err := <-handler.ErrCh:
+	deviceCode := params["device_code"][0]
```

Comment:
This (and other similar lines) probably needs some guards / bounds-checking to prevent runtime panic, maybe?